cli/js/compiler_api.ts

```diff
@@ -0,0 +1,395 @@
+ * file.
+ *
+ *      const results =  await Deno.transpileOnly({
+ *        "foo.ts": `const foo: string = "foo";`
+ *      });
+ *
+ * @param sources A map where the key is the filename and the value is the text
+ *                to transpile.  The filename is only used in the transpile and
+ *                not resolved, for example to fill in the source name in the
+ *                source map.
+ * @param options An option object of options to send to the compiler. This is
+ *                a subset of ts.CompilerOptions which can be supported by Deno.
+ *                Many of the options related to type checking and emitting
+ *                type declaration files will have no impact on the output.
+ */
+export function transpileOnly(
+  sources: Record<string, string>,
+  options?: CompilerOptions
+): Promise<Record<string, TranspileOnlyResult>> {
+  util.log("Deno.transpileOnly", { sources: Object.keys(sources), options });
+  const payload = {
+    sources,
+    options: options ? JSON.stringify(options) : undefined
+  };
+  return sendAsync(dispatch.OP_TRANSPILE, payload).then(result =>
+    JSON.parse(result)
+  );
+}
+
+/** Takes a root module name, any optionally a record set of sources. Resolves
+ * with a compiled set of modules.  If just a root name is provided, the modules
+ * will be resolved as if the root module had been passed on the command line.
+ *
+ * If sources are passed, all modules will be resolved out of this object, where
+ * the key is the module name and the value is the content.  The extension of
+ * the module name will be used to determine the media type of the module.
+ *
+ *      const [ maybeDiagnostics1, output1 ] = await Deno.compile("foo.ts");
+ *
+ *      const [ maybeDiagnostics2, output2 ] = await Deno.compile("/foo.ts", {
+ *        "/foo.ts": `export * from "./bar.ts";`,
+ *        "/bar.ts": `export const bar = "bar";`
+ *      });
+ *
+ * @param rootName The root name of the module which will be used as the
+ *                 "starting point".  If no `sources` is specified, Deno will
+ *                 resolve the module externally as if the `rootName` had been
+ *                 specified on the command line.
+ * @param sources An optional key/value map of sources to be used when resolving
+ *                modules, where the key is the module name, and the value is
+ *                the source content.  The extension of the key will determine
+ *                the media type of the file when processing.  If supplied,
+ *                Deno will not attempt to resolve any modules externally.
+ * @param options An optional object of options to send to the compiler. This is
+ *                a subset of ts.CompilerOptions which can be supported by Deno.
+ */
+export function compile(
+  rootName: string,
+  sources?: Record<string, string>,
+  options?: CompilerOptions
+): Promise<[Diagnostic[] | undefined, Record<string, string>]> {
```

Comment:
Diagnostics are null when there are none, so what would empty array diagnostics mean?

Reply:
Are you getting an empty array?  I believe I coded it so that it doesn't resolve with an empty array.

Reply:
Why that kind of API? Empty array for no diagnostics follows naturally and is just as easy to check for. Unless this is some convention?

Reply:
> Empty array for no diagnostics follows naturally and is just as easy to check for.

Is it?  Technically it is slightly slower, because property access is slower than a value check.

> Unless this is some convention?

There are two conventions.  One is the arguments to a callback convention, where `err` would be the first argument and `undefined` if there was no error.  The second convention is the Rust enum `Result` type, where you have `Err` and `Ok`.  While not directly translatable, a presence of a value would indicate that something didn't go ok.

So why would an empty array be preferable?

Reply:
> There are two conventions. One is the arguments to a callback convention, where `err` would be the first argument and `undefined` if there was no error.

When there is a _single_ potential error, the no-error case is modelled with nullability. When there is (explicitly) a list of errors, the no-error case is modelled naturally as an empty list and nullability is redundant, isn't it? `null | non-empty list` is generally strange to see.

My first instinct compares it more to an `stderr` stream which is simply empty or not (especially when it's referred to as a diagnostic).